### PR TITLE
SAI Acl meta data support

### PR DESCRIPTION
**Pull request by ramangupta**

### CODE REVIEW

File: sai/inc/saifdb.h [hunk 1]
```diff
@@ -94,6 +94,10 @@ typedef enum _sai_fdb_entry_attr_t
     /* FDB entry packet action [sai_packet_action_t] (MANDATORY_ON_CREATE|CREATE_AND_SET) */
     SAI_FDB_ENTRY_ATTR_PACKET_ACTION,
 
+    /* User based Meta Data to qualify FDB parameters in ACL 
+     * [sai_uint32_t] (CREATE_AND_SET) */
+    SAI_FDB_ENTRY_ATTR_ACL_META_DATA,
```
Comment (lguohan) [inline sai/inc/saifdb.h]: default value?

Comment (ramangupta) [inline sai/inc/saifdb.h]: This wouldn't be a mandatory attribute and whenever used should be within the range of values supported by FDB.

File: sai/inc/saiacl.h [hunk 2]
```diff
@@ -367,8 +401,52 @@ typedef enum _sai_acl_entry_attr_t
...
+
+    /* DST MAC address match in FDB 
+     * Value must be in the range defined in
+     * [0 - SAI_SWITCH_ATTR_ACL_FDB_USER_META_DATA_MAXIMUM] */
+    SAI_ACL_ENTRY_ATTR_FIELD_USER_META_FDB_DST,
+
+    /* DST IP address match in L3 Table 
+     * Value must be in the range defined in
+     * [0 - SAI_SWITCH_ATTR_ACL_L3_USER_META_DATA_MAXIMUM] */
+    SAI_ACL_ENTRY_ATTR_FIELD_USER_META_L3_DST,
```
Comment (lguohan) [inline sai/inc/saiacl.h]: rename to SAI_ACL_ENTRY_ATTR_FIELD_USER_META_ROUTE_DST to the naming in sairoute?

File: sai/inc/saineighbor.h [hunk 3]
```diff
@@ -53,6 +53,11 @@ typedef enum _sai_neighbor_attr_t
      * [bool] (CREATE_AND_SET) (default to false) */
     SAI_NEIGHBOR_ATTR_NO_HOST_ROUTE,
 
+    /* User based Meta Data to qualify Neighbor Table parameters in ACL 
+     * [sai_uint32_t] (CREATE_AND_SET) 
+     * Value Range [0 - SAI_SWITCH_ATTR_ACL_L3_USER_META_DATA_MAXIMUM] */
+    SAI_NEIGHBOR_ATTR_ACL_META_DATA,
```
Comment (lguohan) [inline sai/inc/saineighbor.h]: I do not see this to be present in acl match table/entry.

File: sai/inc/saiswitch.h [hunk 4]
```diff
@@ -167,6 +167,21 @@ typedef enum _sai_switch_attr_t
     /* maximum priority for ACL entry [sai_uint32_t] */
     SAI_SWITCH_ATTR_ACL_ENTRY_MAXIMUM_PRIORITY,
 
+    /* maximum FDB user-based ACL meta data [sai_uint32_t] */
+    SAI_SWITCH_ATTR_ACL_FDB_USER_META_DATA_MAXIMUM,
+
+    /* maximum L3 Table user-based ACL meta data [sai_uint32_t] */
+    SAI_SWITCH_ATTR_ACL_L3_USER_META_DATA_MAXIMUM,
```
Comment (lguohan) [inline sai/inc/saiswitch.h]: SAI_SWITCH_ATTR_ACL_ROUTE_USER_META_DATA_MAXIMUM

File: sai/inc/saiacl.h (hunk 2 above)
Comment (ramangupta) [inline sai/inc/saiacl.h]: L3 includes both Route and Neighbor table and one meta data would be applicable for both the tables.

File: sai/inc/saiacl.h [hunk 5]
```diff
@@ -233,8 +233,42 @@ typedef enum _sai_acl_table_attr_t
...
+    /* Meta Data carried from previous ACL Stage */ 
+    SAI_ACL_TABLE_ATTR_FIELD_USER_META_ACL_LOOKUP, 
+
+    /* NPU Based Meta Data [bool] */ 
+ 
+    /* SRC MAC address match in FDB */ 
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_FDB_SRC_HIT, 
+ 
+    /* SRC MAC address matches static FDB entry */ 
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_FDB_SRC_STATIC_HIT,
```
Comment (lguohan) [inline sai/inc/saiacl.h]: remove src match since fdb src match is not defined yet.

File: sai/inc/saiacl.h [hunk 6]
```diff
@@ -233,8 +233,42 @@ typedef enum _sai_acl_table_attr_t
...
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_FDB_SRC_HIT, 
+ 
+    /* SRC MAC address matches static FDB entry */ 
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_FDB_SRC_STATIC_HIT, 
+ 
+    /* DST MAC address match in FDB */ 
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_FDB_DST_HIT, 
+
+    /* SRC IP address match in L3 Table */ 
+    SAI_ACL_TABLE_ATTR_FIELD_NPU_META_L3_SRC_HIT,
```
Comment (lguohan) [inline sai/inc/saiacl.h]: remove l3 src hit as it is not defined yet

File: sai/inc/saiacl.h (hunk 2 above)
Comment (lguohan) [inline sai/inc/saiacl.h]: say on some asics, those two tables are combined and a route entry will point to a next hop entry, and have a 2 byte meta data. In the model, the neighbor table will be map to the next hop entry. Then, you can split the 2 byte meta data and let route meta to use a few bits and neighbor entry to use other bits.

For the acl entry to match, you can use mask to match either route meta data portion or neighbor meta data portion.

File: sai/inc/saiacl.h [hunk 7]
```diff
@@ -367,8 +401,52 @@ typedef enum _sai_acl_entry_attr_t
...
+    SAI_ACL_ENTRY_ATTR_FIELD_NPU_META_FDB_SRC_HIT,
+
+    /* SRC MAC address matches static FDB entry */
+    SAI_ACL_ENTRY_ATTR_FIELD_NPU_META_FDB_SRC_STATIC_HIT,
+
+    /* DST MAC address match in FDB */
+    SAI_ACL_ENTRY_ATTR_FIELD_NPU_META_FDB_DST_HIT,
+
+    /* SRC IP address match in L3 Table */
+    SAI_ACL_ENTRY_ATTR_FIELD_NPU_META_L3_SRC_HIT,
```
Comment (lguohan) [inline sai/inc/saiacl.h]: remove this one.

**Pull request closed without merging**